ksql-serde/src/main/java/io/confluent/ksql/serde/avro/KsqlGenericRowAvroSerializer.java

```diff
@@ -46,48 +51,55 @@
   ByteArrayOutputStream output;
   Encoder encoder;
   List<Schema.Field> fields;
+  private final boolean isInternal;
 
-  public KsqlGenericRowAvroSerializer(org.apache.kafka.connect.data.Schema schema) {
-    this.schema = schema;
-  }
+  KafkaAvroSerializer kafkaAvroSerializer;
 
-  @Override
-  public void configure(final Map<String, ?> map, final boolean b) {
-    rowSchema = (String) map.get(AVRO_SERDE_SCHEMA_CONFIG);
-    if (rowSchema == null) {
-      throw new SerializationException("Avro schema is not set for the serializer.");
+  public KsqlGenericRowAvroSerializer(org.apache.kafka.connect.data.Schema schema,
+                                      SchemaRegistryClient schemaRegistryClient, KsqlConfig
+                                          ksqlConfig, boolean isInternal) {
+    this.isInternal = isInternal;
+    if (isInternal) {
+      this.schema = SchemaUtil.getAvroSerdeKsqlSchema(schema);
+    } else {
+      this.schema = SchemaUtil.getSchemaWithNoAlias(schema);
     }
+
+    String avroSchemaStr = SchemaUtil.buildAvroSchema(schema, "avro_schema");
     parser = new Schema.Parser();
-    avroSchema = parser.parse(rowSchema);
+    avroSchema = parser.parse(avroSchemaStr);
     fields = avroSchema.getFields();
     writer = new GenericDatumWriter<>(avroSchema);
+    Map map = new HashMap();
+    map.put(AbstractKafkaAvroSerDeConfig.AUTO_REGISTER_SCHEMAS, true);
+    map.put(AbstractKafkaAvroSerDeConfig.SCHEMA_REGISTRY_URL_CONFIG, ksqlConfig.getString(KsqlConfig.SCHEMA_REGISTRY_URL_PROPERTY));
+    kafkaAvroSerializer = new KafkaAvroSerializer(schemaRegistryClient, map);
+
+  }
+
+  @Override
+  public void configure(final Map<String, ?> map, final boolean b) {
+
   }
 
   @Override
   public byte[] serialize(final String topic, final GenericRow genericRow) {
     if (genericRow == null) {
       return null;
     }
-    GenericRecord avroRecord = new GenericData.Record(avroSchema);
-    for (int i = 0; i < genericRow.getColumns().size(); i++) {
-      if (fields.get(i).schema().getType() == Schema.Type.ARRAY) {
-        avroRecord.put(fields.get(i).name(), Arrays.asList((Object[]) genericRow.getColumns().get(i)));
-      } else {
-        avroRecord.put(fields.get(i).name(), genericRow.getColumns().get(i));
-      }
-
-    }
-
     try {
-      output = new ByteArrayOutputStream();
-      encoder = EncoderFactory.get().binaryEncoder(output, null);
-      writer.write(avroRecord, encoder);
-      encoder.flush();
-      output.flush();
-    } catch (IOException e) {
-      throw new SerializationException("Error serializing AVRO message", e);
+      GenericRecord avroRecord = new GenericData.Record(avroSchema);
+      for (int i = 0; i < genericRow.getColumns().size(); i++) {
+        if (fields.get(i).schema().getType() == Schema.Type.ARRAY) {
```

Comment:
We don't need special treatments for map?

Reply:
No, map is dealt with as HashMap and works fine when using it in AvroRecord.